test/polygon_reduce.cpp

```diff
@@ -0,0 +1,480 @@
+  return (1e-6 <= u && u <= (1 - 1e-6) && 1e-6 <= t && t <= (1 - 1e-6));
+}
+
+bool safeToRemove(Polygons &polys, int i, int j) {
+  if (polys[i].size() == 3) return false;
+  int prev = j == 0 ? polys[i].size() - 1 : (j - 1);
+  int next = j == (polys[i].size() - 1) ? 0 : (j + 1);
+  for (int k = 0; k < polys.size(); k++) {
+    if (!all_of(ExecutionPolicy::Par, countAt(0lu), countAt(polys[k].size()),
+                [&](int l) {
+                  int ll = l == (polys[k].size() - 1) ? 0 : (l + 1);
+                  if (i == k && (l == j || ll == j)) return true;
+                  return !intersect(polys[i][prev], polys[i][next], polys[k][l],
+                                    polys[k][ll]);
+                }))
+      return false;
+  }
+  return true;
+}
+
+std::pair<int, int> findIndex(const Polygons &polys, int i) {
+  int outer = 0;
+  while (i > polys[outer].size()) i -= polys[outer++].size();
+  return std::make_pair(outer, i);
+}
+
+void Dump(const Polygons &polys) {
+  for (auto poly : polys) {
+    std::cout << "polys.push_back({" << std::setprecision(9) << std::endl;
+    for (auto v : poly) {
+      std::cout << "    {" << v.x << ", " << v.y << "},  //" << std::endl;
+    }
+    std::cout << "});" << std::endl;
+  }
+  for (auto poly : polys) {
+    std::cout << "show(array([" << std::endl;
+    for (auto v : poly) {
+      std::cout << "  [" << v.x << ", " << v.y << "]," << std::endl;
+    }
+    std::cout << "]))" << std::endl;
+  }
+}
+
+void DumpTriangulation(const Polygons &polys) {
+  ExecutionParams oldParams = PolygonParams();
+  manifold::PolygonParams().processOverlaps = true;
+  auto result = Triangulate(polys);
+  PolygonParams() = oldParams;
+  for (auto &tri : result) {
+    auto x = findIndex(polys, tri.x);
+    auto y = findIndex(polys, tri.y);
+    auto z = findIndex(polys, tri.z);
+    printf("show(array([[%.7f, %.7f], [%.7f, %.7f], [%.7f, %.7f]]))\n",
+           polys[x.first][x.second].x, polys[x.first][x.second].y,
+           polys[y.first][y.second].x, polys[y.first][y.second].y,
+           polys[z.first][z.second].x, polys[z.first][z.second].y);
+  }
+}
+
+bool triangulationValid(Polygons &polys, std::vector<glm::ivec3> triangles) {
+  std::vector<std::pair<glm::vec2, glm::vec2>> edges(3 * triangles.size());
+  for (auto &tri : triangles) {
+    auto x = findIndex(polys, tri.x);
+    auto y = findIndex(polys, tri.y);
+    auto z = findIndex(polys, tri.z);
+    edges.push_back(
+        std::make_pair(polys[x.first][x.second], polys[y.first][y.second]));
+    edges.push_back(
+        std::make_pair(polys[y.first][y.second], polys[z.first][z.second]));
+    edges.push_back(
+        std::make_pair(polys[z.first][z.second], polys[x.first][x.second]));
+  }
+  return std::all_of(edges.begin(), edges.end(), [&](auto &p) {
+    return all_of(ExecutionPolicy::Par, edges.begin(), edges.end(),
+                  [&](auto &q) {
+                    return !intersect(p.first, p.second, q.first, q.second);
```

Comment:
For consistency with our Triangulator, you might want to copy my `CheckGeometry` function, which I believe is equivalent, but faster (since we already know the triangulation is manifold).

Reply:
It seems that they are not equivalent. The `CheckGeometry` function checks if the polygons are in CCW orientation, but does not check if they can have intersection.

Reply:
And CCW check seems unnecessary as removing vertices will not invert orientation.